### PR TITLE
save.db to use a mmap file - Dan Molik <dan@d3fy.net>

### PR DESCRIPTION
**Pull request by GrayTShirt**
so it compiles, but fails testing, bolospy seems to be crashing.
at this point I think I need a little help.

### CODE REVIEW

**Comment by GrayTShirt** [comment]
so sleeping on it helped, it turns out I had an off-by-one error in the read function. and now I'm failing large-payload checks, I think what's happening is that I'm trying to copy chunks bigger than pagesize and I'm getting a bus error, I'm going to try chunking the buffer and seeing if that helps.

**Comment by GrayTShirt** [comment]
and updated

File: src/binf.c [hunk 1]
```diff
@@ -82,7 +83,7 @@ static uint64_t ntohll(uint64_t x)
 	return h.u64;
 }
 
-static int s_write_record(int fd, ssize_t *len, uint8_t type, void *_)
+static void s_write_record(void *addr, uint8_t type, void *_, size_t *len)
```
Comment (filefrog) [inline src/binf.c]: Why did you move `size_t *len` to be the last argument?

File: src/binf.c [hunk 2]
```diff
@@ -136,19 +136,14 @@ static int s_write_record(int fd, ssize_t *len, uint8_t type, void *_)
 		break;
 
 	default:
-		return -1;
```
Comment (filefrog) [inline src/binf.c]: How do we signal errors here?

File: src/binf.c [hunk 3]
```diff
@@ -136,19 +136,14 @@ static int s_write_record(int fd, ssize_t *len, uint8_t type, void *_)
...
-
 	record.len   = htons(record.len);
 	record.flags = htons(type);
 
-	n = write(fd, &record, sizeof(record));
-	so_far += n;
-	if (n != sizeof(record))
-		return n;
+	memcpy(addr + *len, &record, sizeof(record));
+	*len += sizeof(record);
```
Comment (filefrog) [inline src/binf.c]: I feel like this two-liner can be expressed as a small macro that only exists for the lexical scope of this function, to clarify the intent and minimize stupid typo bugs...

File: src/binf.c [hunk 4]
```diff
@@ -156,25 +151,16 @@ static int s_write_record(int fd, ssize_t *len, uint8_t type, void *_)
 		body.state.status    = payload.state->status;
 		body.state.stale     = payload.state->stale;
 
-		n = write(fd, &body.state, want = sizeof(body.state));
-		so_far += n;
-		if (n != want) logger(LOG_CRIT, "wanted %i / got %i", want, n);
-		if (n != want)
-			return so_far;
+		memcpy(addr + *len, &body.state, sizeof(body.state));
+		*len += sizeof(body.state);
```
Comment (filefrog) [inline src/binf.c]: ... because it's used here

File: src/binf.c [hunk 5]
```diff
@@ -156,25 +151,16 @@ static int s_write_record(int fd, ssize_t *len, uint8_t type, void *_)
...
 
 		s = payload.state->name;
-		n = write(fd, s, want = strlen(s) + 1);
-		so_far += n;
-		if (n != want) logger(LOG_CRIT, "wanted %i / got %i", want, n);
-		if (n != want)
-			return so_far;
+		memcpy(addr + *len, s, strlen(s) + 1);
+		*len += strlen(s) + 1;
```
Comment (filefrog) [inline src/binf.c]: ... and here

File: src/binf.c [hunk 6]
```diff
@@ -156,25 +151,16 @@ static int s_write_record(int fd, ssize_t *len, uint8_t type, void *_)
...
 
 		s = payload.state->summary;
-		n = write(fd, s, want = strlen(s) + 1);
-		so_far += n;
-		if (n != want) logger(LOG_CRIT, "wanted %i / got %i", want, n);
-		if (n != want)
-			return so_far;
+		memcpy(addr + *len, s, strlen(s) + 1);
+		*len += strlen(s) + 1;
```
Comment (filefrog) [inline src/binf.c]: (etc.)

File: src/binf.c [hunk 7]
```diff
@@ -496,10 +421,12 @@ int binf_write(db_t *db, const char *file)
 	rate_t   *rate;
 
 	char *name;
-	int i, n;
-	ssize_t len;
+	void *addr;
+	int i;
+	size_t *so_far = vmalloc(sizeof(size_t));
+		*so_far = 0;
```
Comment (filefrog) [inline src/binf.c]: `vmalloc` is implemented in terms of `calloc`, so the memory is already (quiet efficiently) zeroed for you.

File: src/binf.c [hunk 8]
```diff
@@ -522,84 +449,65 @@ int binf_write(db_t *db, const char *file)
 	for_each_key_value(&db->rates,    name, rate)    header.count++;
 	header.count = htonl(header.count);
 
-	logger(LOG_INFO, "writing %i bytes for the binary file header", sizeof(header));
-	n = write(fd, &header, sizeof(header));
-	if (n != sizeof(header)) {
-		logger(LOG_ERR, "only wrote %i of %i bytes for the savefile header; %s is probably corrupt now",
-			n, sizeof(header), file);
+	/* allocate 128 bytes per stat plus a page of padding */
+	// size_t fsize = header.count * 128 + sysconf(_SC_PAGE_SIZE);
```
Comment (filefrog) [inline src/binf.c]: Should this just be removed outright?

File: src/binf.c [hunk 9]
```diff
@@ -738,14 +650,39 @@ int binf_read(db_t *db, const char *file)
...
 	char trailer[2] = { 1 };
-	n = read(fd, trailer, 2);
-	if (n != 2 || trailer[0] || trailer[1]) {
+	memcpy(trailer, addr + *so_far, 2);
+	if (trailer[0] || trailer[1]) {
 		logger(LOG_ERR, "no savefile trailer found!");
 		close(fd);
 		return 1;
 	}
+	free(so_far);
```
Comment (filefrog) [inline src/binf.c]: Again, do we need to heap-alloc this since we are freeing it before returning (i.e. it is not a chunk of memory that does/should outlive its defining scope)?

File: src/bolo.h [hunk 10]
```diff
@@ -219,8 +223,9 @@ typedef struct {
 
 #define probable(f) (rand() * 1.0 / RAND_MAX <= (f))
 
-int binf_write(db_t *db, const char *file);
-int binf_read(db_t *db, const char *file);
+int binf_write(db_t *db, const char *file, int db_size);
+int binf_read(db_t *db, const char *file, int db_size);
+int binf_sync(const char *file, int db_size);
```
Comment (filefrog) [inline src/bolo.h]: What is the purpose of `db_size`?

And what does `binf_sync` do?

Can you document these somewhere, either as a block comment (`/* ... */` style) or a standalone file explaining how the binary format behaves?

File: src/config.c [hunk 11]
```diff
@@ -199,6 +201,8 @@ static int lex(parser_t *p)
 			KEYWORD("use",        USE);
 			KEYWORD("max.events", MAXEVENTS);
 			KEYWORD("grace.period", GRACE_PERIOD);
+			KEYWORD("save.size",    SAVE_SIZE);
+			KEYWORD("save.period",  SAVE_PERIOD);
```
Comment (filefrog) [inline src/config.c]: The verbiage "save period" seems clumsy to me, esp. since this is a system dealing with time-series data.  Can we do "save.every" or "save.interval" or "save.frequency" instead?

File: src/config.c [hunk 12]
```diff
@@ -199,6 +201,8 @@ static int lex(parser_t *p)
 			KEYWORD("use",        USE);
 			KEYWORD("max.events", MAXEVENTS);
 			KEYWORD("grace.period", GRACE_PERIOD);
+			KEYWORD("save.size",    SAVE_SIZE);
```
Comment (filefrog) [inline src/config.c]: Why does this need to be tunable, and can we document how / why?

File: src/sub_rrd.c [hunk 13]
```diff
@@ -514,12 +514,12 @@ static int _creator_reactor(void *socket, pdu_t *pdu, void *_) /* {{{ */
 #endif
 
 				rrd_clear_error();
-				if (CACHED && ! rrdc_is_connected(CACHED)) {
-					rc = rrd_create_r(file, 60, time_s() - 365 * 86400,
-						args->num, (const char **)args->strings);
-				} else {
+				if (CACHED && rrdc_is_connected(CACHED)) {
```
Comment (filefrog) [inline src/sub_rrd.c]: This looks like a mostly unrelated change.  Can you split this into a separate PR?

File: t/large-payloads [hunk 14]
```diff
@@ -64,7 +64,8 @@ RATE|$TS|$(payload 16 r1small)|42
 RATE|$TS|$(payload 5100 r2large)|42
 EOF
 
-sleep 1
+sleep 3
+
```
Comment (filefrog) [inline t/large-payloads]: Why did you need to bump the sleep value?

**Comment by filefrog** [comment]
Overall, I think I follow this PR, but I would like to see more documentation, some rationale for changing function signatures (why `addr` is not next to `len` in some of the function parameter lists), loss of error reporting, etc.

Also, split out the RRD stuff into a separate PR.

**Comment by GrayTShirt** [comment]
So I made the changes and deployed them locally;

1. The server is under a lot less stress from constantly writing now.
2. The vmem has skyrocketed from mmap, the map address either needs to be properly deallocated or moved into core.
3. the rrd/map in sub_rrd should probably get the same treatment.

**Comment by filefrog** [comment]
I'm confused - is it writing *more* now than before the mmap changes?

**Comment by GrayTShirt** [comment]
ugh, I really need to proof read these comments.

The changes reduced the load on the system significantly. and I'm no longer getting insert errors in the rrd sub. Which seemed to be a symptom of high IO usage in the core.

**Comment by filefrog** [comment]
Awesome. So that's a vote to merge?

**Comment by GrayTShirt** [comment]
It certainly could be merged as it's working in service, but 2 things should be considered:
1. A smarter way of pre-allocating memory for x-large state messages; perhaps a bloom filter that takes a random sampling, and allocates memory with a wide margin of error? I think the db size should probably be saved in the db-header.
2. a defined method of flushing to disk, IE: do we write to the mmap every 60 seconds, and force a flush after N map write, do we have a timeout, if we haven't committed to disk after Y seconds we force a flush?

One more thing:
is memcpy absolutely necessary in this case, is there a more efficient method?
